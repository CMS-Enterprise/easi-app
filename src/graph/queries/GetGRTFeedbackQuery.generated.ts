import * as Types from '../generated/generated';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions =  {}
export type GrtFeedbackFieldsFragment = (
  { __typename?: 'GRTFeedback' }
  & Pick<Types.GrtFeedback, 'id' | 'feedbackType' | 'feedback' | 'createdAt'>
);

export type GetGrtFeedbackQueryVariables = Types.Exact<{
  intakeID: Types.Scalars['UUID'];
}>;


export type GetGrtFeedbackQuery = (
  { __typename?: 'Query' }
  & { systemIntake?: Types.Maybe<(
    { __typename?: 'SystemIntake' }
    & { grtFeedbacks: Array<(
      { __typename?: 'GRTFeedback' }
      & GrtFeedbackFieldsFragment
    )> }
  )> }
);

export const GrtFeedbackFieldsFragmentDoc = gql`
    fragment GRTFeedbackFields on GRTFeedback {
  id
  feedbackType
  feedback
  createdAt
}
    `;
export const GetGrtFeedbackDocument = gql`
    query GetGRTFeedback($intakeID: UUID!) {
  systemIntake(id: $intakeID) {
    grtFeedbacks {
      ...GRTFeedbackFields
    }
  }
}
    ${GrtFeedbackFieldsFragmentDoc}`;

/**
 * __useGetGrtFeedbackQuery__
 *
 * To run a query within a React component, call `useGetGrtFeedbackQuery` and pass it any options that fit your needs.
 * When your component renders, `useGetGrtFeedbackQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useGetGrtFeedbackQuery({
 *   variables: {
 *      intakeID: // value for 'intakeID'
 *   },
 * });
 */
export function useGetGrtFeedbackQuery(baseOptions: Apollo.QueryHookOptions<GetGrtFeedbackQuery, GetGrtFeedbackQueryVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<GetGrtFeedbackQuery, GetGrtFeedbackQueryVariables>(GetGrtFeedbackDocument, options);
      }
export function useGetGrtFeedbackLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<GetGrtFeedbackQuery, GetGrtFeedbackQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useLazyQuery<GetGrtFeedbackQuery, GetGrtFeedbackQueryVariables>(GetGrtFeedbackDocument, options);
        }
export type GetGrtFeedbackQueryHookResult = ReturnType<typeof useGetGrtFeedbackQuery>;
export type GetGrtFeedbackLazyQueryHookResult = ReturnType<typeof useGetGrtFeedbackLazyQuery>;
export type GetGrtFeedbackQueryResult = Apollo.QueryResult<GetGrtFeedbackQuery, GetGrtFeedbackQueryVariables>;