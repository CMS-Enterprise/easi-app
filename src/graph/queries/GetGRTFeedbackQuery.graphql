fragment GRTFeedbackFields on GRTFeedback {
  id
  feedbackType
  feedback
  createdAt
}

query GetGRTFeedback($intakeID: UUID!) {
  systemIntake(id: $intakeID) {
    grtFeedbacks {
      ...GRTFeedbackFields
    }
  }
}