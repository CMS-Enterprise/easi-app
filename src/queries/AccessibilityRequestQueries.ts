import { gql } from '@apollo/client';

export const CreateAccessibilityRequest = gql`
  mutation CreateAccessibilityRequest(
    $input: CreateAccessibilityRequestInput!
  ) {
    createAccessibilityRequest(input: $input) {
      accessibilityRequest {
        id
        name
      }
      userErrors {
        message
        path
      }
    }
  }
`;

export const DeleteAccessibilityRequest = gql`
  mutation DeleteAccessibilityRequest(
    $input: DeleteAccessibilityRequestInput!
  ) {
    deleteAccessibilityRequest(input: $input) {
      id
    }
  }
`;
