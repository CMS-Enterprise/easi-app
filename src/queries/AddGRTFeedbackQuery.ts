import { gql } from '@apollo/client';

export default gql`
  mutation AddGRTFeedback($input: AddGRTFeedbackInput!) {
    addGRTFeedback(input: $input) {
      id
    }
  }
`;
