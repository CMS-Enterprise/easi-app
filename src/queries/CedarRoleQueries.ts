import { gql } from '@apollo/client';

const CedarRoleType = gql`
  fragment CedarRoleType on CedarRoleType {
    id
    name
    description
  }
`;

export const GetCedarRoleTypesQuery = gql`
  ${CedarRoleType}
  query GetCedarRoleTypes {
    roleTypes {
      ...CedarRoleType
    }
  }
`;

export const SetRolesForUserOnSystemQuery = gql`
  mutation SetRolesForUserOnSystem($input: SetRolesForUserOnSystemInput!) {
    setRolesForUserOnSystem(input: $input)
  }
`;
