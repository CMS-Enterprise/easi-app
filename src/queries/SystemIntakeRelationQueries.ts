import { gql } from '@apollo/client';

export const GetSystemIntakeRelationQuery = gql`
  query GetSystemIntakeRelation($id: UUID!) {
    systemIntake(id: $id) {
      id
      relationType
      contractName
      contractNumbers {
        contractNumber
      }
      systems {
        id
        name
        acronym
      }
    }
    cedarSystems {
      id
      name
    }
  }
`;

export const SetSystemIntakeRelationNewSystemQuery = gql`
  mutation SetSystemIntakeRelationNewSystem(
    $input: SetSystemIntakeRelationNewSystemInput!
  ) {
    setSystemIntakeRelationNewSystem(input: $input) {
      systemIntake {
        id
      }
    }
  }
`;

export const SetSystemIntakeRelationExistingSystemQuery = gql`
  mutation SetSystemIntakeRelationExistingSystem(
    $input: SetSystemIntakeRelationExistingSystemInput!
  ) {
    setSystemIntakeRelationExistingSystem(input: $input) {
      systemIntake {
        id
      }
    }
  }
`;

export const SetSystemIntakeRelationExistingServiceQuery = gql`
  mutation SetSystemIntakeRelationExistingService(
    $input: SetSystemIntakeRelationExistingServiceInput!
  ) {
    setSystemIntakeRelationExistingService(input: $input) {
      systemIntake {
        id
      }
    }
  }
`;

export const UnlinkSystemIntakeRelationQuery = gql`
  mutation UnlinkSystemIntakeRelation($intakeID: UUID!) {
    unlinkSystemIntakeRelation(intakeID: $intakeID) {
      systemIntake {
        id
      }
    }
  }
`;

export const GetTrbRequestRelationQuery = gql`
  query GetTrbRequestRelation($id: UUID!) {
    trbRequest(id: $id) {
      id
      relationType
      contractName
      contractNumbers {
        contractNumber
      }
      systems {
        id
        name
        acronym
      }
    }
    cedarSystems {
      id
      name
    }
  }
`;
export const SetTrbRequestRelationNewSystemQuery = gql`
  mutation SetTrbRequestRelationNewSystem(
    $input: SetTRBRequestRelationNewSystemInput!
  ) {
    setTRBRequestRelationNewSystem(input: $input) {
      id
    }
  }
`;

export const SetTrbRequestRelationExistingSystemQuery = gql`
  mutation SetTrbRequestRelationExistingSystem(
    $input: SetTRBRequestRelationExistingSystemInput!
  ) {
    setTRBRequestRelationExistingSystem(input: $input) {
      id
    }
  }
`;

export const SetTrbRequestRelationExistingServiceQuery = gql`
  mutation SetTrbRequestRelationExistingService(
    $input: SetTRBRequestRelationExistingServiceInput!
  ) {
    setTRBRequestRelationExistingService(input: $input) {
      id
    }
  }
`;

export const UnlinkTrbRequestRelationQuery = gql`
  mutation UnlinkTrbRequestRelation($trbRequestID: UUID!) {
    unlinkTRBRequestRelation(trbRequestID: $trbRequestID) {
      id
    }
  }
`;
