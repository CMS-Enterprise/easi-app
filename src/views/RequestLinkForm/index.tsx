import React, { useEffect, useMemo, useState } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { Trans, useTranslation } from 'react-i18next';
import { Link, useHistory, useLocation, useParams } from 'react-router-dom';
import { useMutation, useQuery } from '@apollo/client';
import {
  Breadcrumb,
  BreadcrumbBar,
  BreadcrumbLink,
  Button,
  ButtonGroup,
  Fieldset,
  Form,
  FormGroup,
  Grid,
  IconArrowBack,
  Label,
  Radio,
  TextInput
} from '@trussworks/react-uswds';
import classNames from 'classnames';

import MainContent from 'components/MainContent';
import Modal from 'components/Modal';
import PageHeading from 'components/PageHeading';
import PageLoading from 'components/PageLoading';
import Alert from 'components/shared/Alert';
import IconButton from 'components/shared/IconButton';
import MultiSelect from 'components/shared/MultiSelect';
import RequiredAsterisk from 'components/shared/RequiredAsterisk';
import {
  GetSystemIntakeRelationQuery,
  SetSystemIntakeRelationExistingServiceQuery,
  SetSystemIntakeRelationExistingSystemQuery,
  SetSystemIntakeRelationNewSystemQuery,
  UnlinkSystemIntakeRelationQuery
} from 'queries/SystemIntakeRelationQueries';
import {
  GetSystemIntakeRelation,
  GetSystemIntakeRelationVariables
} from 'queries/types/GetSystemIntakeRelation';
import {
  SetSystemIntakeRelationExistingService,
  SetSystemIntakeRelationExistingServiceVariables
} from 'queries/types/SetSystemIntakeRelationExistingService';
import {
  SetSystemIntakeRelationExistingSystem,
  SetSystemIntakeRelationExistingSystemVariables
} from 'queries/types/SetSystemIntakeRelationExistingSystem';
import {
  SetSystemIntakeRelationNewSystem,
  SetSystemIntakeRelationNewSystemVariables
} from 'queries/types/SetSystemIntakeRelationNewSystem';
import {
  UnlinkSystemIntakeRelation,
  UnlinkSystemIntakeRelationVariables
} from 'queries/types/UnlinkSystemIntakeRelation';
import { RequestRelationType } from 'types/graphql-global-types';

const RequestLinkForm = () => {
  const { systemId } = useParams<{
    systemId: string;
  }>();
  const history = useHistory();

  const { t } = useTranslation(['itGov', 'intake', 'action', 'error']);

  // Form mode is either new or edit
  const { state } = useLocation<{ isNew?: boolean }>();
  const isNew = !!state?.isNew;

  // Fetch query param to check if coming from ITGov/TRB admin home for redirect and text changes
  const params = new URLSearchParams(useLocation().search);
  const editType = params.get('edit-type');

  const taskListUrl = `/governance-task-list/${systemId}`;
  let breadCrumb = t('additionalRequestInfo.taskListBreadCrumb');

  let redirectUrl = taskListUrl;
  if (editType === 'it-gov-admin') {
    redirectUrl = `/governance-review-team/${systemId}/additional-information`;
    breadCrumb = t('additionalRequestInfo.itGovBreadcrumb');
  } else if (editType === 'trb-admin') {
    redirectUrl = `/trb/${systemId}/additional-information`;
    breadCrumb = t('additionalRequestInfo.trbBreadcrumb');
  }

  const [hasUserError, setUserError] = useState<boolean>(false);

  const [isSkipModalOpen, setSkipModalOpen] = useState<boolean>(false);
  const [isUnlinkModalOpen, setUnlinkModalOpen] = useState<boolean>(false);

  const { data, error, loading } = useQuery<
    GetSystemIntakeRelation,
    GetSystemIntakeRelationVariables
  >(GetSystemIntakeRelationQuery, { variables: { id: systemId } });

  const cedarSystemIdOptions = useMemo(() => {
    const cedarSystemsData = data?.cedarSystems;
    return !cedarSystemsData
      ? []
      : cedarSystemsData.map(system => ({
          label: system.name,
          value: system.id
        }));
  }, [data?.cedarSystems]);

  const [
    setSystemIntakeRelationNewSystem,
    { error: newSystemError }
  ] = useMutation<
    SetSystemIntakeRelationNewSystem,
    SetSystemIntakeRelationNewSystemVariables
  >(SetSystemIntakeRelationNewSystemQuery);

  const [
    setSystemIntakeRelationExistingSystem,
    { error: existingSystemError }
  ] = useMutation<
    SetSystemIntakeRelationExistingSystem,
    SetSystemIntakeRelationExistingSystemVariables
  >(SetSystemIntakeRelationExistingSystemQuery);

  const [
    setSystemIntakeRelationExistingService,
    { error: existingServiceError }
  ] = useMutation<
    SetSystemIntakeRelationExistingService,
    SetSystemIntakeRelationExistingServiceVariables
  >(SetSystemIntakeRelationExistingServiceQuery);

  const [unlinkSystemIntakeRelation, { error: unlinkError }] = useMutation<
    UnlinkSystemIntakeRelation,
    UnlinkSystemIntakeRelationVariables
  >(UnlinkSystemIntakeRelationQuery);

  const { control, watch, setValue, handleSubmit } = useForm<{
    relationType: RequestRelationType | null;
    cedarSystemIDs: string[];
    contractNumbers: string;
    contractName: string;
  }>({
    defaultValues: {
      relationType: null,
      cedarSystemIDs: [],
      contractNumbers: '',
      contractName: ''
    },
    values: (values => {
      if (!values) return undefined;
      return {
        relationType: values.relationType,
        cedarSystemIDs: values.systems.map(v => v.id),
        contractNumbers: values.contractNumbers
          .map(v => v.contractNumber)
          .join(', '),
        contractName: values.contractName || ''
      };
    })(data?.systemIntake)
  });

  // Ref fields for some form behavior
  const fields = watch();
  const relation = fields.relationType;

  // This form doesn't use field validation feedback
  // Instead the submission button is disabled according to field requirements
  const submitEnabled: boolean = (() => {
    if (relation === null) return false;

    if (relation === RequestRelationType.NEW_SYSTEM) return true;

    if (
      relation === RequestRelationType.EXISTING_SYSTEM &&
      fields.cedarSystemIDs.length
    )
      return true;

    if (
      relation === RequestRelationType.EXISTING_SERVICE &&
      fields.contractName.trim() !== ''
    )
      return true;

    // Default to disabled
    return false;
  })();

  const submit = handleSubmit(formData => {
    // Do some field parsing and correlate relation type to mutation

    // Parse contract numbers from csv text input to string[]
    // Make sure an empty string input is sent as an empty list
    const contractNumbers = formData.contractNumbers
      .split(',')
      .map(v => v.trim())
      .filter(v => v !== '');

    let p;

    if (relation === RequestRelationType.NEW_SYSTEM) {
      p = setSystemIntakeRelationNewSystem({
        variables: {
          input: {
            systemIntakeID: systemId,
            contractNumbers
          }
        }
      });
    } else if (relation === RequestRelationType.EXISTING_SYSTEM) {
      p = setSystemIntakeRelationExistingSystem({
        variables: {
          input: {
            systemIntakeID: systemId,
            cedarSystemIDs: formData.cedarSystemIDs,
            contractNumbers
          }
        }
      });
    } else if (relation === RequestRelationType.EXISTING_SERVICE) {
      p = setSystemIntakeRelationExistingService({
        variables: {
          input: {
            systemIntakeID: systemId,
            contractName: formData.contractName,
            contractNumbers
          }
        }
      });
    }

    p?.then(
      res => {
        if (res?.data) history.push(redirectUrl);
      },
      () => {}
    ).catch(() => {
      setUserError(true);
    });
  });

  const unlink = () => {
    unlinkSystemIntakeRelation({ variables: { intakeID: systemId } })
      .then(
        res => {
          if (res?.data) history.push(redirectUrl);
        },
        () => {}
      )
      .catch(() => {
        setUserError(true);
      });
  };

  // Error feedback
  const hasErrors =
    error ||
    existingServiceError ||
    existingSystemError ||
    newSystemError ||
    unlinkError ||
    hasUserError;

  useEffect(() => {
    if (hasErrors) {
      const err = document.getElementById('link-form-error');
      err?.scrollIntoView();
    }
  }, [hasErrors]);

  return (
    <MainContent className="grid-container margin-bottom-15">
      {hasErrors && (
        <Alert id="link-form-error" type="error" slim className="margin-top-2">
          {t('error:encounteredIssueTryAgain')}
        </Alert>
      )}
      {loading && <PageLoading />}
      {!loading && data && (
        <>
          <BreadcrumbBar variant="wrap">
            <Breadcrumb>
              <BreadcrumbLink asCustom={Link} to="/">
                <span>{t('intake:navigation.itGovernance')}</span>
              </BreadcrumbLink>
            </Breadcrumb>
            {isNew ? (
              <Breadcrumb current>
                {t('intake:navigation.startRequest')}
              </Breadcrumb>
            ) : (
              <>
                <Breadcrumb>
                  <BreadcrumbLink asCustom={Link} to={redirectUrl}>
                    <span>{breadCrumb}</span>
                  </BreadcrumbLink>
                </Breadcrumb>
                <Breadcrumb current>
                  {t('intake:navigation.editLinkRelation')}
                </Breadcrumb>
              </>
            )}
          </BreadcrumbBar>
          <PageHeading className="margin-top-4 margin-bottom-0">
            {t('link.header')}
          </PageHeading>
          <p className="font-body-lg line-height-body-5 text-light margin-y-0">
            {t('link.description')}
          </p>
          <p className="margin-top-2 margin-bottom-5 text-base">
            <Trans
              i18nKey="action:fieldsMarkedRequired"
              components={{ asterisk: <RequiredAsterisk /> }}
            />
          </p>

          <Form
            className="easi-form maxw-full"
            onSubmit={e => e.preventDefault()}
          >
            <Grid row>
              <Grid tablet={{ col: 12 }} desktop={{ col: 6 }}>
                <Fieldset
                  legend={
                    <h4 className="margin-top-0 margin-bottom-1 line-height-heading-2">
                      {t('link.form.field.systemOrService.label')}
                    </h4>
                  }
                >
                  {/* New system or service */}
                  <Radio
                    id="relationType-newSystem"
                    name="relationType"
                    value={RequestRelationType.NEW_SYSTEM}
                    label={t('link.form.field.systemOrService.options.0')}
                    onChange={() =>
                      setValue('relationType', RequestRelationType.NEW_SYSTEM)
                    }
                    checked={relation === RequestRelationType.NEW_SYSTEM}
                  />

                  {relation === RequestRelationType.NEW_SYSTEM && (
                    <Controller
                      name="contractNumbers"
                      control={control}
                      render={({ field }) => (
                        <FormGroup className="margin-left-4">
                          <Label
                            htmlFor="contractNumber"
                            hint={t('link.form.field.contractNumberNew.help')}
                          >
                            {t('link.form.field.contractNumberNew.label')}
                          </Label>
                          <TextInput
                            {...field}
                            ref={null}
                            id="contractNumbers"
                            type="text"
                          />
                        </FormGroup>
                      )}
                    />
                  )}

                  {/* Existing system */}
                  <Radio
                    id="relationType-existingSystem"
                    name="relationType"
                    value={RequestRelationType.EXISTING_SYSTEM}
                    label={t('link.form.field.systemOrService.options.1')}
                    onChange={() =>
                      setValue(
                        'relationType',
                        RequestRelationType.EXISTING_SYSTEM
                      )
                    }
                    checked={relation === RequestRelationType.EXISTING_SYSTEM}
                  />

                  {relation === RequestRelationType.EXISTING_SYSTEM && (
                    <div className="margin-left-4">
                      <Controller
                        name="cedarSystemIDs"
                        control={control}
                        render={({ field }) => (
                          <FormGroup>
                            <Label
                              htmlFor="cedarSystemIDs"
                              hint={t('link.form.field.cmsSystem.help')}
                            >
                              {t('link.form.field.cmsSystem.label')}{' '}
                              <RequiredAsterisk />
                            </Label>
                            <MultiSelect
                              name={field.name}
                              selectedLabel={t(
                                'link.form.field.cmsSystem.selectedLabel'
                              )}
                              initialValues={field.value}
                              options={cedarSystemIdOptions}
                              onChange={values => field.onChange(values)}
                            />
                          </FormGroup>
                        )}
                      />

                      <Controller
                        name="contractNumbers"
                        control={control}
                        render={({ field }) => (
                          <FormGroup>
                            <Label
                              htmlFor="contractNumber"
                              hint={t(
                                'link.form.field.contractNumberExisting.help'
                              )}
                            >
                              {t(
                                'link.form.field.contractNumberExisting.label'
                              )}
                            </Label>
                            <TextInput
                              {...field}
                              ref={null}
                              id="contractNumbers"
                              type="text"
                            />
                          </FormGroup>
                        )}
                      />
                    </div>
                  )}

                  {/* Existing service or contract */}
                  <Radio
                    id="relationType-existingService"
                    name="relationType"
                    value={RequestRelationType.EXISTING_SERVICE}
                    label={t('link.form.field.systemOrService.options.2')}
                    onChange={() =>
                      setValue(
                        'relationType',
                        RequestRelationType.EXISTING_SERVICE
                      )
                    }
                    checked={relation === RequestRelationType.EXISTING_SERVICE}
                  />

                  {relation === RequestRelationType.EXISTING_SERVICE && (
                    <div className="margin-left-4">
                      <Controller
                        name="contractName"
                        control={control}
                        render={({ field }) => (
                          <FormGroup>
                            <Label htmlFor="contractName">
                              {t('link.form.field.serviceOrContractName.label')}{' '}
                              <RequiredAsterisk />
                            </Label>
                            <TextInput
                              {...field}
                              ref={null}
                              id="contractName"
                              type="text"
                            />
                          </FormGroup>
                        )}
                      />

                      <Controller
                        name="contractNumbers"
                        control={control}
                        render={({ field }) => (
                          <FormGroup>
                            <Label
                              htmlFor="contractNumber"
                              hint={t(
                                'link.form.field.contractNumberExisting.help'
                              )}
                            >
                              {t(
                                'link.form.field.contractNumberExisting.label'
                              )}
                            </Label>
                            <TextInput
                              {...field}
                              ref={null}
                              id="contractNumbers"
                              type="text"
                            />
                          </FormGroup>
                        )}
                      />
                    </div>
                  )}
                </Fieldset>
              </Grid>
            </Grid>

            <ButtonGroup>
              <Button
                type="submit"
                disabled={!submitEnabled}
                onClick={() => submit()}
              >
                {t(
                  `link.form.${(() => {
                    if (isNew) {
                      return relation === RequestRelationType.NEW_SYSTEM
                        ? 'continueTaskList'
                        : 'next';
                    }
                    return 'saveChanges';
                  })()}`
                )}
              </Button>

              {(isNew ||
                (!isNew && data.systemIntake?.relationType !== null)) && (
                <Button
                  type="submit"
                  unstyled
                  onClick={() => {
                    if (isNew) setSkipModalOpen(true);
                    else setUnlinkModalOpen(true);
                  }}
                  className={classNames('margin-left-1', {
                    'text-error': !isNew
                  })}
                >
                  {t(`link.form.${isNew ? 'skip' : 'unlink'}`)}
                </Button>
              )}
            </ButtonGroup>

            <IconButton
              icon={<IconArrowBack className="margin-right-05" />}
              type="button"
              unstyled
              onClick={() => {
                history.goBack();
              }}
            >
              {t('link.form.back')}
            </IconButton>

            {/* Skip confirm modal */}
            <Modal
              isOpen={isSkipModalOpen}
              closeModal={() => setSkipModalOpen(false)}
            >
              <h2 className="usa-modal__heading margin-bottom-2">
                {t('link.skipConfirm.heading')}
              </h2>
              <p className="margin-y-0">{t('link.skipConfirm.text')}</p>
              <ul className="easi-list margin-top-0">
                <li>{t('link.skipConfirm.list.0')}</li>
                <li>{t('link.skipConfirm.list.1')}</li>
              </ul>
              <ButtonGroup className="margin-top-3">
                <Button type="button" onClick={() => history.push(taskListUrl)}>
                  {t('link.skipConfirm.submit')}
                </Button>
                <Button
                  type="button"
                  unstyled
                  className="margin-left-1"
                  onClick={() => setSkipModalOpen(false)}
                >
                  {t('link.skipConfirm.cancel')}
                </Button>
              </ButtonGroup>
            </Modal>

            {/* Unlink confirm */}
            <Modal
              isOpen={isUnlinkModalOpen}
              closeModal={() => setUnlinkModalOpen(false)}
            >
              <h2 className="usa-modal__heading margin-bottom-2">
                {t('link.unlinkConfirm.heading')}
              </h2>
              <p className="margin-top-0">{t('link.unlinkConfirm.text.0')}</p>
              <p className="margin-bottom-0">
                {t('link.unlinkConfirm.text.1')}
              </p>
              <ul className="easi-list margin-top-0">
                <li>{t('link.unlinkConfirm.list.0')}</li>
                <li>{t('link.unlinkConfirm.list.1')}</li>
              </ul>
              <ButtonGroup className="margin-top-3">
                <Button secondary type="button" onClick={() => unlink()}>
                  {t('link.unlinkConfirm.submit')}
                </Button>
                <Button
                  type="button"
                  unstyled
                  className="margin-left-1"
                  onClick={() => setUnlinkModalOpen(false)}
                >
                  {t('link.unlinkConfirm.cancel')}
                </Button>
              </ButtonGroup>
            </Modal>
          </Form>
        </>
      )}
    </MainContent>
  );
};

export default RequestLinkForm;
