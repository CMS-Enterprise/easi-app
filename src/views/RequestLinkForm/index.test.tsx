import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import {
  render,
  screen,
  waitForElementToBeRemoved
} from '@testing-library/react';

import { GetSystemIntakeRelationQuery } from 'queries/SystemIntakeRelationQueries';
import VerboseMockedProvider from 'utils/testing/VerboseMockedProvider';

import RequestLinkForm from '.';

describe('IT Gov Request relation link form', () => {
  it('renders', async () => {
    const id = 'fa93173c-2e8c-4371-b464-4b3dd649f940';
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/system/link/${id}`]}>
        <VerboseMockedProvider
          mocks={[
            {
              request: {
                query: GetSystemIntakeRelationQuery,
                variables: {
                  id
                }
              },
              result: {
                data: {
                  systemIntake: {
                    id,
                    relationType: null,
                    contractName: null,
                    contractNumbers: [],
                    systems: [],
                    __typename: 'SystemIntake'
                  },
                  cedarSystems: [
                    {
                      id: '{11AB1A00-1234-5678-ABC1-1A001B00CC0A}',
                      name: 'Centers for Management Services',
                      __typename: 'CedarSystem'
                    }
                  ]
                }
              }
            }
          ]}
          addTypename={false}
        >
          <Route path="/system/link/:systemId?">
            <RequestLinkForm />
          </Route>
        </VerboseMockedProvider>
      </MemoryRouter>
    );

    await waitForElementToBeRemoved(() => screen.getByTestId('page-loading'));

    expect(asFragment()).toMatchSnapshot();
  });
});
