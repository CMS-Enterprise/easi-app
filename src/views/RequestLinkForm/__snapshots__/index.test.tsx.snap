// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IT Gov Request relation link form > renders 1`] = `
<DocumentFragment>
  <main
    class="easi-main-content grid-container margin-bottom-15"
    id="main-content"
    role="main"
    tabindex="-1"
  >
    <nav
      aria-label="Breadcrumbs"
      class="usa-breadcrumb usa-breadcrumb--wrap"
    >
      <ol
        class="usa-breadcrumb__list"
      >
        <li
          class="usa-breadcrumb__list-item"
        >
          <a
            class="usa-breadcrumb__link"
            href="/"
          >
            <span>
              IT Governance
            </span>
          </a>
        </li>
        <li
          class="usa-breadcrumb__list-item"
        >
          <a
            class="usa-breadcrumb__link"
            href="/governance-task-list/fa93173c-2e8c-4371-b464-4b3dd649f940"
          >
            <span>
              Task list
            </span>
          </a>
        </li>
        <li
          aria-current="page"
          class="usa-breadcrumb__list-item usa-current"
        >
          Edit linked system, service, or contract
        </li>
      </ol>
    </nav>
    <h1
      aria-live="polite"
      class="easi-h1 margin-top-4 margin-bottom-0"
      tabindex="-1"
    >
      Is this request part of an existing system, service, or contract?
    </h1>
    <p
      class="font-body-lg line-height-body-5 text-light margin-y-0"
    >
      If you are requesting a Life Cycle ID (LCID) and governance approval for a new IT investment that is a part of an existing system, service, or contract (such as a major change based on a new business need or the creation of a new sub-system) please include additional details below.
    </p>
    <p
      class="margin-top-2 margin-bottom-5 text-base"
    >
      Fields marked with an asterisk (
      <span
        class="text-red margin-x-05"
      >
        *
      </span>
      ) are required.
    </p>
    <form
      class="usa-form easi-form maxw-full"
      data-testid="form"
    >
      <div
        class="grid-row"
        data-testid="grid"
      >
        <div
          class="tablet:grid-col-12 desktop:grid-col-6"
          data-testid="grid"
        >
          <fieldset
            class="usa-fieldset"
            data-testid="fieldset"
          >
            <legend
              class="usa-legend"
            >
              <h4
                class="margin-top-0 margin-bottom-1 line-height-heading-2"
              >
                Is this IT Governance request related to an existing system, service, or contract?
              </h4>
            </legend>
            <div
              class="usa-radio"
              data-testid="radio"
            >
              <input
                class="usa-radio__input"
                id="relationType-newSystem"
                name="relationType"
                type="radio"
                value="NEW_SYSTEM"
              />
              <label
                class="usa-radio__label"
                for="relationType-newSystem"
              >
                No, this is a completely new system, service, or contract
              </label>
            </div>
            <div
              class="usa-radio"
              data-testid="radio"
            >
              <input
                class="usa-radio__input"
                id="relationType-existingSystem"
                name="relationType"
                type="radio"
                value="EXISTING_SYSTEM"
              />
              <label
                class="usa-radio__label"
                for="relationType-existingSystem"
              >
                Yes, an existing system
              </label>
            </div>
            <div
              class="usa-radio"
              data-testid="radio"
            >
              <input
                class="usa-radio__input"
                id="relationType-existingService"
                name="relationType"
                type="radio"
                value="EXISTING_SERVICE"
              />
              <label
                class="usa-radio__label"
                for="relationType-existingService"
              >
                Yes, an existing service or other contract
              </label>
            </div>
          </fieldset>
        </div>
      </div>
      <ul
        class="usa-button-group"
      >
        <li
          class="usa-button-group__item"
        >
          <button
            class="usa-button"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Save changes
          </button>
        </li>
        <li
          class="usa-button-group__item"
        />
      </ul>
      <button
        class="usa-button usa-button--unstyled display-flex flex-align-center"
        data-testid="button"
        type="button"
      >
        <svg
          class="usa-icon margin-right-05 margin-right-1"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
        Back
      </button>
    </form>
  </main>
</DocumentFragment>
`;
